### PR TITLE
Strong agent

### PR DESCRIPTION
**Pull request by pahautelman**
Strong agent:
- can do every action
- can carry two blocks at the same time
-  note: changed the `max_objects` param in `BW4TBrain` file as suggested [here](https://stackoverflow.com/c/tud-cs/a/11857/826)

`find_action` method ranks possible actions in the following order:
1. if next goal block, `searching_next` has been located, and the agent in not carrying another goal block, pick it
2.  if next goal block has been located and the (manhattan) distance to `searching_next` is smaller than the distance to the drop location, grab the  `searching_next` block
3. if you're carrying the current goal block, deliver it
4. if current goal block has been located, go pick it up
5. find random room that none of the agents searched through
6. find random room that the agent has not searched
7. explore random room

TODO
- [x] merge branch `main` into `strong_agent`
- [ ] add smarter heuristic for grabbing other objects
- [ ] select closest door instead of random door
- [ ] test how it works with multiple agents
- [ ] update `phase` to `None` after every tick when delivering a goal block or moving to a room, so that the agent can select the best action they can do

### CODE REVIEW

File: agents1/Message.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    room_name = re.findall("(room_\d)", content)
+    return room_name[0]
+
+
+class MessageBuilder:
+
+    def __init__(self, agent_name):
+        self.agent_name = agent_name
+
+    def create_message(self, mt, room_name=None, block_vis=None, location=None):
```
Comment (kmariunas) [inline agents1/Message.py]: We should probably have an option to send a private message to some agent

**Review by andreasbien — COMMENTED** [review]
For the strong agent I would implement a functionality that allows the agent to pick it's next goal based on how many goal blocks there are in a place. (This can be used as a heuristic, to let the strong agent pick up 2 blocks instead of 1) I will work on it if you let me once it's refactored

**Comment by pahautelman** [comment]
> For the strong agent I would implement a functionality that allows the agent to pick it's next goal based on how many goal blocks there are in a place. (This can be used as a heuristic, to let the strong agent pick up 2 blocks instead of 1) I will work on it if you let me once it's refactored

for sure, corndog! will let you know once its done 🍃🍃

File: agents1/StrongAgent.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+            # pick it up if you're not carrying anything already
+            if len(self._is_carrying) == 0:
+                self._grab_block = searching_next
+                return Phase.PLAN_PATH_TO_BLOCK
+
+            # if you're only carrying one other block, aka you're delivering it already
+            if len(self._is_carrying) == 1:
+                # pick this block up only if it is closer than the goal location
+                agent_loc = state[self.agent_name]['location']
+                if manhattan_dist(agent_loc, self._goal_blocks[searching_next]['location']) \
```
Comment (kmariunas) [inline agents1/StrongAgent.py:36]: Even if this is False, if the strong agent has to pick up the other block, we would actually lose ticks. This is potentially an edge case, but we should keep it in mind

**Review by kmariunas — APPROVED** [review]
Very nice

**Review by andreasbien — COMMENTED** [review]
Thank you for your amazing work! I will implement the algorithm stuff now!

**Comment by andreasbien** [comment]
Hmm, there seems to be a bug in the code. 
image

Never mind
I realized I did something wrong. Disregard this message

**Pull request merged**